### PR TITLE
hcm: handle stream reset during downstream encoding

### PR DESCRIPTION
**Pull request by danzh2010**
Signed-off-by: Dan Zhang <danzh@google.com>

Commit Message: when the downstream codec encodes response header/body/trailers in the same call stack as upstream decoding, which is the case for Http3, a connection could gets closed which will cause the stream to be reset. In such case, FilterManagerImpl shouldn't continue process the rest of the response, especially end_stream.

Additional Description: this contributes to the crash in #19006. This fix stops the crash, but didn't make the test pass.

Risk Level: low
Testing: new unit tests
Release Notes: guarded by envoy.reloadable_features.handle_stream_reset_during_hcm_encoding

### CODE REVIEW

**Review by alyssawilk — COMMENTED** [review]
I'd definitely like to see this integration tested.
I believe we can enable DISABLED_AdminDrainDrainsListeners with this change, but also with some potential improvements to get the new HandleSocketFail test to work for HTTP/3?

**Review by alyssawilk — COMMENTED** [review]
Also one high level question for @mattklein123 

The problem here is we're writing to a closed socket, which causes us to close the connection, and reset the stream under the stack of a write, and then we were processing end stream.

I wonder if instead of immediate connection close, we'd be better off scheduling the close for the next loop to avoid closing the downstream connection under the stack of the upstream connection?

**Comment by danzh2010** [comment]
> Also one high level question for @mattklein123
> 
> The problem here is we're writing to a closed socket, which causes us to close the connection, and reset the stream under the stack of a write, and then we were processing end stream.
> 
> I wonder if instead of immediate connection close, we'd be better off scheduling the close for the next loop to avoid closing the downstream connection under the stack of the upstream connection?

In this case, the connection was closed by QUICHE in its write call stack. I don't think we can delay the close. What we can delay is HCM's reset callback.

**Comment by mattklein123** [comment]
> I wonder if instead of immediate connection close, we'd be better off scheduling the close for the next loop to avoid closing the downstream connection under the stack of the upstream connection?

No objection from me at a high level if we need to do deferred processing, though I haven't spent the time to look at this in detail. If you want me to look at this change in more detail I can spend some time tomorrow.

**Comment by alyssawilk** [comment]
nah I buy dan's argument that it's not worth the refactor
/wait on the integration test now my PR landed

**Comment by danzh2010** [comment]
> 

Integration test is added. PTAL

File: test/integration/protocol_integration_test.cc
Comment (alyssawilk) [inline test/integration/protocol_integration_test.cc:3559]: remove TODO?

**Review by alyssawilk — COMMENTED** [review]
lgtm modulo one remaining nit

File: test/integration/protocol_integration_test.cc
Comment (danzh2010) [reply test/integration/protocol_integration_test.cc]: done

**Review by alyssawilk — APPROVED** [review]

**Review by snowp — APPROVED** [review]
LGTM, thanks!

**Comment by danzh2010** [comment]
Can it be merged?

**Pull request merged**
